Apply hilbert stc

BabaSanfour opened the pull request:
Fixes #12310.


Updating mne.filter.MixinFilter methods to work with Source Estimated data

BabaSanfour commented:
working to fix the failed tests

BabaSanfour commented:
All tests passed.
I think the functions are decent. 
For SourceEstimate.resample the function overwrite the MixinFilter.resample so I did not update the latter. However I think that SourceEstimate.resample can be integrated into MixinFilter.resample. Not sure if this would have any implication on speed but I think it would improve the code.

larsoner reviewed (commented):
Looks like a good start! See some inline comments.

Any new functionality should be accompanied by unit tests, so in `mne/tests/test_source_estimate.py` we'll need some basic tests of the new functions that have been added for at least the `SourceEstimate` class.

BabaSanfour commented:
I have addressed all of the suggestions. I also added one function to test to `stc.apply_function` in `test_source_estimate.py` . For the `object.apply_hilbert` function, the `test_evoked.py` has the test function for raw, epochs and evoked objects, I wonder if the tests should be added in the same function. Also, to test `object.apply_function` there is pre-existing data in `../io/tests/data` . Should we use for example raw data to generate sources? or is there a way to get access for stc_test_data ?

larsoner commented on mne/filter.py:
```diff
@@ -2515,6 +2515,8 @@ def savgol_filter(self, h_freq, verbose=None):
 
         .. versionadded:: 0.9.0
 
+        When working on SourceEstimates the sample rate of the original data is inferred from tstep.
+
```
Versionadded always goes at the bottom
```suggestion
        When working on SourceEstimates the sample rate of the original data is inferred from tstep.

        .. versionadded:: 0.9.0
```

larsoner commented on mne/filter.py:
```diff
@@ -2622,6 +2630,9 @@ def filter(
                   ``len(picks) * n_times`` additional time points need to
                   be temporarily stored in memory.
 
+        .. note:: When working on SourceEstimates the sample rate of the original
+                    data is inferred from tstep.
+
```
Probably not critical enough to be a `.. note::`
```suggestion
        When working on SourceEstimates the sample rate of the original
        data is inferred from tstep.

```

larsoner commented on mne/utils/docs.py:
```diff
@@ -1550,6 +1551,9 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
     " because it will apply channel-wise"
 )
+docdict["fun_applyfun_stc"] = applyfun_fun_base.format(
+    " because it will apply vertices-wise"
```
```suggestion
    " because it will apply vertex-wise"
```

larsoner reviewed (commented):
> For the object.apply_hilbert function, the test_evoked.py has the test function for raw, epochs and evoked objects, I wonder if the tests should be added in the same function.

Sure!

> Should we use for example raw data to generate sources? or is there a way to get access for stc_test_data ?

I would just use `np.random.default_rng(0).normal(...)` like here:

https://github.com/mne-tools/mne-python/blob/e6b49ea773d522670a534ec4a03470ab9c2baf01/mne/tests/test_source_estimate.py#L373-L379

But don't use `np.random.seed`, that's bad practice -- use class-based methods after instantiating a `default_rng` instead (modern best practice for NumPy random that we haven't fully adopted yet but should at some point!).

larsoner commented on mne/tests/test_source_estimate.py:
```diff
@@ -2054,3 +2054,31 @@ def test_label_extraction_subject(kind):
+    vertices = [np.array(np.arange(50)), np.array(np.arange(50, 100))]
+    tmin = 0.0
+    tstep = 0.001
+    data = np.random.rand(n_vertices, n_times)
```
```suggestion
    data = np.random.default_rng(0).normal(size=(n_vertices, n_times))
```

larsoner commented on mne/tests/test_source_estimate.py:
```diff
@@ -2054,3 +2054,31 @@ def test_label_extraction_subject(kind):
+    stc_copy = stc.copy()
+    stc.apply_function(fun)
+    for idx in range(n_vertices):
+        assert_array_equal(stc._data[idx, :], 2 * stc_copy._data[idx, :])
```
Always prefer public attributes when possible, and for float arithmetic it's a good idea to use `assert_allclose` rather than `assert_array_equal`
```suggestion
        assert_allclose(stc.data[idx, :], 2 * stc_copy.data[idx, :])
```

larsoner commented on mne/tests/test_source_estimate.py:
```diff
@@ -2054,3 +2054,31 @@ def test_label_extraction_subject(kind):
+    # Test applying the function with parallelization
+    stc.apply_function(fun, n_jobs=2)
+    for idx in range(n_vertices):
+        assert_array_equal(stc._data[idx, :], 4 * stc_copy._data[idx, :])
```
```suggestion
        assert_allclose(stc.data[idx, :], 4 * stc_copy.data[idx, :])
```

larsoner reviewed (approved):
Just three minor tweaks, marking for merge-when-green, thanks in advance @BabaSanfour !
